### ISSUE TITLE
Rule proposal: NoTrailingWhitespace

### ISSUE THREAD

**Issue by AArnott** [DotNetAnalyzers/StyleCopAnalyzers#729]
As found [here](https://github.com/AArnott/nerdbank.stylecop.rules/blob/master/src/NerdBank.StyleCop.Rules/NerdBankRules.cs) and described [here](http://blog.nerdbank.net/2008/09/notrailingwhitespace-stylecop-rule-and.html), the NoTrailingWhitespace rule helps avoid code where there is invisible trailing whitespace at the end of a line of code. 

Git and some other tools will sometimes flag when there is trailing whitespace, so this helps detect it early.

This issue tracks porting this rule to StyleCop.Analyzers. Feel free to assign to me.

**Comment by sharwell** [comment]
I have this written for Roslyn already btw. :wink:

**Comment by AArnott** [comment]
Oh, awesome. it's just not added to StyleCop.Analyzers then?

**Comment by sharwell** [comment]
That's correct.

**Comment by AArnott** [comment]
ah, well certainly feel free to do it yourself, or point me at your code. No need to re-invent the wheel. :)

**Comment by sharwell** [comment]
Untested, and may use old versions of the Roslyn API. I encourage you to update it to match the form of the other StyleCop analyzers if you decide to use it. :+1: 

```cs
namespace WhitespaceDiagnostic
{
    using System.Collections.Immutable;
    using System.Linq;
    using Microsoft.CodeAnalysis;
    using Microsoft.CodeAnalysis.CSharp;
    using Microsoft.CodeAnalysis.Diagnostics;

    [DiagnosticAnalyzer(LanguageNames.CSharp)]
    public class WhitespaceDiagnosticAnalyzer : DiagnosticAnalyzer
    {
        public const string DiagnosticId = "WhitespaceDiagnostic";
        internal const string Title = "Code must not contain trailing whitespace";
        internal const string MessageFormat = "Remove trailing whitespace";
        internal const string Category = "Style";

        internal static DiagnosticDescriptor Rule = new DiagnosticDescriptor(DiagnosticId, Title, MessageFormat, Category, DiagnosticSeverity.Warning, isEnabledByDefault: true, customTags: new[] { WellKnownDiagnosticTags.Unnecessary });

        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics
        {
            get
            {
                return ImmutableArray.Create(Rule);
            }
        }

        public override void Initialize(AnalysisContext context)
        {
            context.RegisterSyntaxTreeAction(HandleSyntaxTree);
        }

        /// <summary>
        /// 
        /// </summary>
        /// <param name="context"></param>
        private void HandleSyntaxTree(SyntaxTreeAnalysisContext context)
        {
            var root = context.Tree.GetRoot(context.CancellationToken);
            foreach (var trivia in root.DescendantTrivia(descendIntoTrivia: true))
            {
                switch (trivia.CSharpKind())
                {
                case SyntaxKind.WhitespaceTrivia:
                    break;

                default:
                    continue;
                }

                bool reportWarning = false;
                var token = trivia.Token;
                SyntaxTriviaList triviaList;
                if (token.LeadingTrivia.Contains(trivia))
                {
                    triviaList = token.LeadingTrivia;
                }
                else
                {
                    triviaList = token.TrailingTrivia;
                }

                bool foundWhitespace = false;
                foreach (var innerTrivia in triviaList)
                {
                    if (!foundWhitespace)
                    {
                        if (innerTrivia.Equals(trivia))
                            foundWhitespace = true;

                        continue;
                    }

                    if (innerTrivia.IsKind(SyntaxKind.EndOfLineTrivia))
                        reportWarning = true;

                    break;
                }

                if (reportWarning)
                {
                    context.ReportDiagnostic(Diagnostic.Create(Rule, trivia.GetLocation()));
                }
            }
        }
    }
}
```

Code fix provider:

```cs
namespace WhitespaceDiagnostic
{
    using System.Collections.Immutable;
    using System.Composition;
    using System.Threading.Tasks;
    using Microsoft.CodeAnalysis;
    using Microsoft.CodeAnalysis.CodeActions;
    using Microsoft.CodeAnalysis.CodeFixes;
    using Microsoft.CodeAnalysis.CSharp;

    [ExportCodeFixProvider("WhitespaceDiagnosticCodeFixProvider", LanguageNames.CSharp), Shared]
    public class WhitespaceDiagnosticCodeFixProvider : CodeFixProvider
    {
        public sealed override ImmutableArray<string> GetFixableDiagnosticIds()
        {
            return ImmutableArray.Create(WhitespaceDiagnosticAnalyzer.DiagnosticId);
        }

        public sealed override FixAllProvider GetFixAllProvider()
        {
            return WellKnownFixAllProviders.BatchFixer;
        }

        public sealed override async Task ComputeFixesAsync(CodeFixContext context)
        {
            foreach (var diagnostic in context.Diagnostics)
            {
                var root = await context.Document.GetSyntaxRootAsync(context.CancellationToken).ConfigureAwait(false);

                var diagnosticSpan = diagnostic.Location.SourceSpan;
                var trivia = root.FindTrivia(diagnosticSpan.Start, findInsideTrivia: true);
                var newRoot = root.ReplaceTrivia(trivia, new SyntaxTrivia[0]);

                var newDocument = context.Document.WithSyntaxRoot(newRoot);

                context.RegisterFix(CodeAction.Create("Remove trailing whitespace", newDocument), diagnostic);
            }
        }
    }
}
```
